src/Catalyst.Core.Modules.Web3/Web3EthApiExtensions.cs

```diff
@@ -0,0 +1,194 @@
+                SenderAddress = transactionCall.From.Bytes.ToByteString(),
+                ReceiverAddress = transactionCall.To?.Bytes.ToByteString() ?? ByteString.Empty,
+                GasLimit = (ulong) transactionCall.Gas.GetValueOrDefault(),
+                GasPrice = transactionCall.GasPrice.GetValueOrDefault().ToUint256ByteString(),
+                Amount = transactionCall.Value.GetValueOrDefault().ToUint256ByteString(), 
+                Data = transactionCall.Data?.ToByteString() ?? ByteString.Empty,
+                Timestamp = Timestamp.FromDateTime(DateTime.UtcNow) //TODO: this should be set by user and signed
+            };
+        }
+        
+        public static TransactionForRpc ToTransactionForRpc(this IWeb3EthApi api, DeltaWithCid deltaWithCid, int transactionIndex)
+        {
+            var (delta, deltaCid) = deltaWithCid;
+            var publicEntry = delta.PublicEntries[transactionIndex];
+            var deltaNumber = delta.DeltaNumber;
+
+            return new TransactionForRpc
+            {
+                GasPrice = publicEntry.GasPrice.ToUInt256(),
+                BlockHash = deltaCid,
+                BlockNumber = (UInt256) deltaNumber,
+                Nonce = publicEntry.Nonce,
+                To = ToAddress(publicEntry.ReceiverAddress),
+                From = ToAddress(publicEntry.SenderAddress),
+                Value = publicEntry.Amount.ToUInt256(),
+                Hash = publicEntry.GetHash(api.HashProvider),
+                Data = publicEntry.Data.ToByteArray(),
+                R = new byte[0],
+                S = new byte[0],
+                V = UInt256.Zero,
+                Gas = publicEntry.GasLimit,
+                TransactionIndex = (UInt256) transactionIndex
+            };
+        }
+
+        static Address ToAddress(ByteString address)
+        {
+            if (address == null || address.IsEmpty)
+            {
+                return null;
+            }
+
+            return new Address(address.ToByteArray());
+        }
+
+        public static CallOutputTracer CallAndRestore(this IWeb3EthApi api, TransactionForRpc transactionCall, DeltaWithCid deltaWithCid)
+        {
+            var parentDelta = deltaWithCid.Delta;
+            Keccak root = parentDelta.StateRoot.ToKeccak();
+
+            if (transactionCall.Gas == null)
+            {
+                transactionCall.Gas = parentDelta.GasLimit;
+            }
+
+            var publicEntry = api.ToPublicEntry(transactionCall, root);
+
+            var newDelta = deltaWithCid.CreateOneOffDelta(publicEntry);
+
+            CallOutputTracer callOutputTracer = new CallOutputTracer();
+
+            api.StateProvider.StateRoot = root;
+            api.Executor.CallAndReset(newDelta, callOutputTracer);
+            api.StateProvider.Reset();
+            api.StorageProvider.Reset();
+            return callOutputTracer;
+        }
+    }
+
+    public struct DeltaWithCid
+    {
+        public Delta Delta;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4979576)